aptos-move/aptos-vm-types/src/module_and_script_storage/state_view_adapter.rs

```diff
@@ -0,0 +1,546 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::module_and_script_storage::{
+    module_storage::AptosModuleStorage, script_storage::AptosScriptStorage,
+    state_view_adapter::ModuleStorageEntry::Deserialized,
+};
+use aptos_types::{
+    on_chain_config::{Features, OnChainConfig},
+    state_store::{state_key::StateKey, StateView},
+    vm::configs::aptos_prod_deserializer_config,
+};
+use bytes::Bytes;
+#[cfg(test)]
+use claims::{assert_matches, assert_ok};
+use move_binary_format::{
+    deserializer::DeserializerConfig,
+    errors::{PartialVMError, PartialVMResult},
+    file_format::CompiledScript,
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    metadata::Metadata,
+    vm_status::StatusCode,
+};
+use move_vm_runtime::{script_hash, Module, ModuleStorage, Script, ScriptStorage};
+use std::{
+    cell::RefCell,
+    collections::{btree_map, hash_map, BTreeMap, HashMap},
+    sync::Arc,
+};
+
+macro_rules! module_storage_error {
+    ($addr:ident, $name:ident, $err:ident) => {
+        PartialVMError::new(StatusCode::STORAGE_ERROR).with_message(format!(
+            "Unexpected storage error for module {}::{}: {:?}",
+            $addr, $name, $err
+        ))
+    };
+}
+
+macro_rules! module_linker_error {
+    ($addr:ident, $name:ident) => {
+        PartialVMError::new(StatusCode::LINKER_ERROR)
+            .with_message(format!("Module {}::{} does not exist", $addr, $name))
+    };
+}
+
+#[derive(Debug)]
+enum ModuleStorageEntry {
+    Deserialized(Arc<CompiledModule>),
+    Verified(Arc<Module>),
+}
+
+#[derive(Debug)]
+enum ScriptStorageEntry {
+    Deserialized(Arc<CompiledScript>),
+    Verified(Arc<Script>),
+}
+
+/// A simple not thread-safe implementation of code storage on top of a state view.
+/// It is never built directly by clients - only via [AsAptosCodeStorage] trait.
```

Comment:
Is this struct only intended for testing purpose?

Reply:
No, also debugger, unit tests, API, anything that wants its StateView to be treated as code cache.